Arduino_related_code/ValveCaliberationTest.py

```diff
@@ -0,0 +1,52 @@
+import tkinter as tk
+from threading import Thread
+from Arduino_related_code.ArduinoDigital import ArduinoDigital
+from Arduino_related_code.ValveControl import ValveControl
+import time
+
+def calibrate_valve(pin, duration_ms, arduino):
+    valve_control = ValveControl(pin, arduino)
+    duration_s = duration_ms / 1000  # Convert duration from milliseconds to seconds
+
+    for i in range(25):
+        print(f"Calibrating valve on pin {pin}: Iteration {i+1}")
+        valve_control.OpenValve(duration_s)  # Open valve for the specified duration in seconds
+
+        while valve_control.IsValveOpen():
+            time.sleep(0.1)  # Check every 0.1 seconds
+
+        print(f"Valve on pin {pin} closed.")
+        time.sleep(1)  # Short delay between iterations
+
+# Function to handle calibration in a separate thread
+def start_calibration(pin, duration_ms):
+    thread = Thread(target=calibrate_valve, args=(pin, duration_ms, arduino))
```

Comment:
Anushka I know you like Threads but they really are a cause for headache. Use them only if you need too and there is no need for them here. You can run the calibration in the TK thread. it will make no difference.

Reply:
ok